Ticket 113: update call status

andioneto opened the pull request:
**New Service:**
- New service explicitly for updating status on call model (updateCallStatus

**Controller Changes:**
- 400 response if: status !== 'ASSIGNED' || status !== 'IN_PROGRESS' || status !== 'HUNG_UP'
- 400 response for invalid status transtion [ allowed: assigned -> in_prog, in_prog -> hung_up
- If new status is 'ATTEMPTED', will check for responses + outcome, throws 404 if none

**Future TODO:**
- will need to modify once ATTEMPTED is switched over to done

andioneto committed:
over loaded the record attempt controller to cover status updates as well
created a new service to update status only

next: state transition validations for controller

andioneto committed:
deleted checkCallIsAssigned before I realized that the logic to release a call used it, added it back.

next: update testing

for now, WIP PR

andioneto committed:
added testing for the new service
removed an extraneous file from csv helper test

perfectlynormalbeast commented on server/controllers/calls.js:
```diff
@@ -25,13 +57,23 @@ function outcomeIsValid(outcome) {
-  const outcomeMap = {
-    DO_NOT_CALL: updateContactDoNotCallById,
-    BAD_NUMBER: updateContactInvalidNumberById
+function validateStatus(currStatus, prevStatus) {
```
Perhaps rename this to `validateStatusForUpdate` to be more specific that other statuses are valid, just that we don't support directly updating to them?

andioneto replied:
✔️

perfectlynormalbeast commented on server/controllers/calls.js:
```diff
@@ -66,42 +108,28 @@ function afterPutCallAttempt(res, outcome, contact_id, attempt_num, campaign_id)
+  // responses will not exist in a status update for HUNG_UP and IN_PROGRESS
+  if (newStatus === 'ATTEMPTED') {
+    if (!responses || !outcome) {
+      res.staus(404).json({ message: 'update request with a status of ATTEMPTED must have response object and outcome string' });
```
400 is a better code. 404 is reserved for when things aren't found/don't exist.

andioneto replied:
Got it ✔️

perfectlynormalbeast reviewed (changes requested):
Some changes requested

perfectlynormalbeast reviewed (changes requested):
Left a few comments.

andioneto committed:
made requested changes for review
changed validateStatus to validateStatusForUpdate
changed 404 status to 400 for invalid status during updateAttempt call

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-update-call-status

andioneto committed:
integrated status param with front end code

now, when you click a button where it used to update status only in the store, it dispatches an action that puts a new status on the call log instance and then updates the status with the new call obj

andioneto committed:
fixed a ci error

perfectlynormalbeast commented on public/src/components/calls/pre_call_btn_grp.jsx:
```diff
@@ -11,8 +11,9 @@ export default class PreCallButtonGroup extends Component {
-    updateCallStatus('IN_PROGRESS');
+    const { call_id, user_id, campaign_id, updateAttempt } = this.props;
+    updateAttempt(user_id, campaign_id, call_id, 'IN_PROGRESS');
+    // updateCallStatus('IN_PROGRESS');
```
Can we drop this line?

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -107,16 +107,21 @@ export function releaseCall(userId, campaignId, callId, currentCallStatus, next
+  .then((currentCall) => {
+    const { call: currentCallObj } = currentCall.data;
+    const { status: currentCallStatus } = currentCallObj;
+    if (currentCallStatus === 'ATTEMPTED') {
```
Hmm maybe let's leave this out? Because it's a hidden side effect? I think a calling method should explicitly call assignToCall (such as in the submit call action).

perfectlynormalbeast replied:
Also, can we update the test for this action?

andioneto replied:
✔️

perfectlynormalbeast reviewed (changes requested):
A few comments

andioneto committed:
debugging bad name, skip, etc.

andioneto committed:
more debugging

andioneto committed:
removed comments
updated updateStatus test
when form submission flow is complete these actions will need to be adjusted again

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-update-call-status

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -78,6 +78,7 @@ export function assignToCall(userId, campaignId) {
     .then((call) => {
       const { data: callObj } = call;
       const { status, outcome } = callObj;
+      console.log('SHOULD BE NEW CALL: ', callObj.id, status);
```
Can we drop this if it was for debugging?

andioneto replied:
lol sorry

perfectlynormalbeast commented on public/src/store.jsx:
```diff
@@ -38,9 +38,9 @@ const middleware = [immutable(), createLogger(), promise(), thunk];
 
 // to reset the store:
-// persistStore(store).purge();
+persistStore(store).purge();
```
Can we switch this around again?

andioneto replied:
yeppers ✔️

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -6,10 +6,6 @@ import ActiveCallControl from './active_call_control';
 export default class CallsSideBar extends Component {
   constructor(props) {
     super(props);
-    const clickHandlers = ['handleStartCallClick', 'handleNextClick', 'handleStopClick'];
```
I'm curious. Why don't we need the changes in this file anymore?

andioneto replied:
After 'ActiveCallControl' was created all of these methods were moved from this file. 
When I was updating click handlers with the new updateAttempt method I noticed that this extra code was still here, so I removed it as an extension of the current cleanup.

perfectlynormalbeast reviewed (changes requested):
Small comments

andioneto committed:
responded to comments and made requested changes:
removed extra console.log
